### PR TITLE
State size limit (#205)

### PR DESCRIPTION
**Pull request by mgbennet**
Per [this](https://github.com/Microsoft/BotFramework-Emulator/issues/205) issue, a configurable limit has been placed on state storage.  It is accessible from a new tab in the App Settings dialog.

### CODE REVIEW

File: src/client/dialogs/appSettingsDialog.tsx [hunk 1]
```diff
@@ -91,7 +105,7 @@ export class AppSettingsDialog extends React.Component<{}, { ngrokPath: string }
             if (filenames && filenames.length) {
                 // TODO: validate selection
                 this.ngrokPathInputRef.value = filenames[0];
-                this.setState({ ngrokPath: filenames[0] });
+                this.setState({ ngrokPath: filenames[0] })
```
Comment (eanders-MS) [inline src/client/dialogs/appSettingsDialog.tsx]: Did you have something against that semi-colon? :)

File: src/server/controllers/emulator/emulatorController.ts [hunk 2]
```diff
@@ -248,7 +248,7 @@ export class EmulatorController {
         windowManager.zoomIn();
     }
     static zoomOut = () => {
-        windowManager.zoomIn();
+        windowManager.zoomOut();
```
Comment (eanders-MS) [inline src/server/controllers/emulator/emulatorController.ts]: Weird, I thought this was already in master.

File: src/utils.ts [hunk 3]
```diff
@@ -126,4 +126,28 @@ export const safeStringify = (o: any, space: string | number = undefined): strin
         }
         return value;
     }, space);
+}
+
+export const approximateObjectSize = (object: any) => {
```
Comment (eanders-MS) [inline src/utils.ts]: What does this return if object is a string or a number?

File: src/server/controllers/emulator/emulatorController.ts (hunk 2 above)
Comment (eanders-MS) [inline src/server/controllers/emulator/emulatorController.ts]: It is in master. What is this comparing against?

Comment (mgbennet) [inline src/server/controllers/emulator/emulatorController.ts]: Yeah, both changes in the last push I made are showing up here, even though they're in master.

File: src/utils.ts (hunk 3 above)
Comment (mgbennet) [inline src/utils.ts]: Strings return 2*(the string length), which is correct, but it does it in a really ineffecient manner. With numbers and booleans it returns zero, which is incorrect. I could fix this pretty easily with a little refactor of the function structure.

File: src/server/controllers/emulator/emulatorController.ts (hunk 2 above)
Comment (eanders-MS) [inline src/server/controllers/emulator/emulatorController.ts]: Maybe try merging master into your branch?

File: src/utils.ts (hunk 3 above)
Comment (eanders-MS) [inline src/utils.ts]: I think you will need to protect against infinite recursion:
```javascript
var obj = { s: "hello" }
obj.t = obj;
approximateObjectSize(obj);
```

Comment (eanders-MS) [inline src/utils.ts]: Take a look at how we do that in safeStringify

**Comment by eanders-MS** [comment]
Changes look good, I will go ahead and merge them, but I have one change request: to move "KB" out of the label and put it next to the number input box:
image
Thanks!

File: src/utils.ts [hunk 4]
```diff
@@ -126,4 +126,31 @@ export const safeStringify = (o: any, space: string | number = undefined): strin
...
+            return 8;
+        case 'string':
+            return object.length * 2;
+        case 'object':
+            let bytes = 0;
+            for (let i in object) {
+                let value = object[i];
+                if (typeof value === 'object' && value !== null) {
+                    if (cache.indexOf(value) !== -1) {
+                        return;
```
Comment (eanders-MS) [inline src/utils.ts]: Oh there's a bug right here. No return value. Should return 0?

Comment (eanders-MS) [inline src/utils.ts]: Actually, I think this should be a break, not a return.

Comment (eanders-MS) [inline src/utils.ts]: Er, make that a continue. Not a break. Not a return. A `continue`.

**Comment by AndreMantas** [comment]
Will this size be configurable in dev.botframework.com for each bot?
I'm having lots of problems with this state size (Error 400 BadRequest in my bot code) and I'm changing my code to fix the problem but it would be really nice to have a bigger limit.

**Comment by eanders-MS** [comment]
State size limit in the Bot Framework is fixed for all bots at 64kb. If this doesn't work for your bot, you can implement your own state service using the BotBuilder-Azure extensions: https://github.com/Microsoft/BotBuilder-Azure. I suggest taking a look at the Samples folder below the CSharp or Node folder (according to your language preference).

More about the state service: https://docs.microsoft.com/bot-framework/dotnet/bot-builder-dotnet-state

**Comment by AndreMantas** [comment]
So in my dialog should I delete all my private fields that I suspect are big and use the Get/Set UserData methods to save and load those big objects?

**Comment by eanders-MS** [comment]
I don't think it will help. All fields are stored in a single data blob, and the max size of that blob is fixed at 64kb. If you cannot pare down your overall state size, I recommend implementing your own state service.

**Comment by AndreMantas** [comment]
So my best bet is to use Azure DocumentDB to store the bot state?

**Comment by eanders-MS** [comment]
Yes, I think that's the right direction, though I haven't done it myself so I don't have much more information than that. If you need more guidance or discussion, you should be able to find it by asking over at https://github.com/Microsoft/BotBuilder/issues or on Stack Overflow (tag with bot framework).

**Pull request merged**